src/browser-connection/set.js

```diff
@@ -0,0 +1,178 @@
+import remove from '../utils/array-remove';
+
+
+function getConst (name, defaultValue) {
+    var value = GLOBAL[name];
+
+    if (value === void 0)
+        return defaultValue;
+
+    return value;
+}
+
+export default class BrowserConnectionSet extends EventEmitter {
+    constructor () {
+        super();
+
+        this.BROWSER_CONNECTION_READY_TIMEOUT = getConst('BROWSER_CONNECTION_READY_TIMEOUT', 30 * 1000);
+        this.WAITING_FOR_IDLE_TIMEOUT         = getConst('WAITING_FOR_IDLE_TIMEOUT', 10000);
+
+        this.connections        = [];
+        this.bindedErrorHandler = this._connectionErrorHandler.bind(this);
+        this.promisedCloses     = [];
+        this.task               = null;
+
+        //NOTE: We're setting an empty error handler, because Node kills the process on an 'error' event
+        //if there is no handler. See https://nodejs.org/api/events.html#events_class_events_eventemitter
+        this.on('error', () => {});
+    }
+
+    _cleanJobQueue (bc) {
+        if (bc instanceof LocalBrowserConnection)
+            bc.jobQueue = [];
+        else {
+            if (!bc.jobQueue.length || !this.task)
+                return;
+
+            if (this.task.pendingBrowserJobs.indexOf(bc.jobQueue[0]) > -1)
+                bc.removeJob(bc.jobQueue[0]);
+        }
+    }
+
+    _createConnectionIdlePromise (bc) {
+        return new Promise(resolve => {
+            if (bc.idle || bc.closed) {
+                resolve();
+                return;
+            }
+            var waitingForIdleTimeout = setTimeout(() => resolve(), this.WAITING_FOR_IDLE_TIMEOUT);
+
+            var eventHandler = () => {
+                clearTimeout(waitingForIdleTimeout);
+                resolve();
+            };
+
+            bc.once('idle', eventHandler);
+            bc.once('closed', eventHandler);
+        });
+    }
+
+    _createConnectionClosePromise (bc) {
+        return new Promise(resolve => {
+            if (bc.closed) {
+                resolve();
+                return;
+            }
+
+            bc.close();
+            bc.once('closed', resolve);
+        });
+    }
+
+    _createConnectionsReadyPromise () {
+        var ready = this.connections
+            .filter(bc => !bc.ready)
+            .map(bc => new Promise(resolve => bc.once('ready', resolve)));
+
+        return Promise.all(ready);
+    }
+
+    async _checkForDisconnectedConnections () {
+        var disconnectedUserAgents = this.connections
+            .filter(bc => bc.closed)
+            .map(bc => bc.userAgent);
+
+        if (disconnectedUserAgents.length) {
+            await this.closeConnections();
+
+            throw new Error(getText(MESSAGE.cantRunAgainstDisconnectedBrowsers, disconnectedUserAgents.join(', ')));
+        }
+    }
+
+    async _connectionErrorHandler (msg) {
+        await this.closeConnections();
+
+        this.emit('error', msg);
+    }
+
+    //API
+    async addConnections (...connections) {
```

Comment:
I don't get this API. Things are still to complicated. We should have just 3 methods:
1)`useConnections(connections)` - adds and waits till connections will be ready, throws error if browsers disconnected, since then it will emit `error` event
2)`freeConnection(connection)` - closes/frees connection
3)`dispose` - closes/frees all connections, also called from `error` event